homeassistant/components/mcp/coordinator.py

```diff
@@ -0,0 +1,79 @@
+"""Types for the Model Context Protocol integration."""
+
+from collections.abc import AsyncGenerator
+from contextlib import AbstractAsyncContextManager, asynccontextmanager
+import datetime
+import logging
+
+import httpx
+from mcp.client.session import ClientSession
+from mcp.client.sse import sse_client
+from mcp.types import Tool
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_URL
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+UPDATE_INTERVAL = datetime.timedelta(minutes=30)
+
+
+@asynccontextmanager
+async def mcp_client(url: str) -> AsyncGenerator[ClientSession]:
+    """Create a server-sent event MCP client.
+
+    This is an asynccontxt manager that wraps to other async context managers
+    so that the coordinator has a single object to manage.
+    """
+    try:
+        async with sse_client(url=url) as streams, ClientSession(*streams) as session:
+            await session.initialize()
+            yield session
+    except ExceptionGroup as err:
+        raise err.exceptions[0] from err
+
+
+class ModelContextProtocolCoordinator(DataUpdateCoordinator[list[Tool]]):
+    """Define an object to hold MCP data."""
+
+    config_entry: ConfigEntry
+    session: ClientSession
+
+    def __init__(self, hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+        """Initialize ModelContextProtocolCoordinator."""
+        super().__init__(
+            hass,
+            logger=_LOGGER,
+            name=DOMAIN,
+            config_entry=config_entry,
+            update_interval=UPDATE_INTERVAL,
+        )
+        self.ctx_mgr: AbstractAsyncContextManager[ClientSession]
+
+    async def _async_setup(self) -> None:
+        """Set up the client connection."""
+        self.ctx_mgr = mcp_client(self.config_entry.data[CONF_URL])
```

Comment:
Is there a better way to manage the lifecycle of a contextmanager client library?

I'm essentially having it try to mange the lifecycle manually here to start/end the context when the coordinator starts/ends.

Reply:
The only alternative I could think of, and I think we do this one for Matter server or Z-Wave server (forgot which):

```python
async def _connect(self, connection_established: asyncio.Event) -> None:
    try:
        async with mcp_client(url) as session:
            self.session = session
            connection_established.set()
    finally:
        self.session = None

async def _async_setup(self) -> None:
    connection_established = asyncio.Event()
    self.config_entry.async_create_background_task(self._connect(connection_established))
    async with timeout(10):
        await connection_established.wait()

```

Reply:
And I guess you can keep the connection task around to cancel it for other reasons as well.

Reply:
Z-Wave similar example https://github.com/home-assistant-libs/zwave-js-server-python/blob/master/zwave_js_server/client.py#L228

Reply:
And that's managed in HA integration:
- https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/__init__.py#L224-L226
- https://github.com/home-assistant/core/blob/dev/homeassistant/components/zwave_js/__init__.py#L926

Reply:
Thanks, updated to follow those examples.